packages/kit/src/core/sync/write_env.js

```diff
@@ -0,0 +1,183 @@
+import { write_if_changed } from './utils.js';
+import path from 'path';
+import fs from 'fs';
+import { loadEnv } from 'vite';
+
+const autogen_comment = '// this section is auto-generated';
+
+/**
+ * @param {boolean} pub
+ * @param {Record<string, string>} env
+ * @returns {string}
+ */
+function const_declaration_template(pub, env) {
+	return `${autogen_comment}
+
+${Object.entries(env)
+	.map(
+		([k, v]) => `/**
+ * @type {import('$app/env${pub ? '' : '/private'}').${k}}
+ */
+export const ${k} = ${JSON.stringify(v)};
+`
+	)
+	.join('\n')}
+`;
+}
+
+/**
+ * @param {boolean} pub
+ * @param {Record<string, string>} env
+ * @returns {string}
+ */
+function type_declaration_template(pub, env) {
+	return `declare module '$app/env${pub ? '' : '/private'}' {
+  ${Object.keys(env)
+		.map((k) => `export const ${k}: string;`)
+		.join('\n  ')}
+}`;
+}
+
+function runtime_env_template() {
+	return `/** @type {App.RuntimeEnv} */
+export let env = {};
+
+/** @type {(environment: Record<string, string>) => void} */
+export function set_env(environment) {
+  env = environment;
+}`;
+}
```

Comment:
this could probably just be a file in https://github.com/sveltejs/kit/tree/master/packages/kit/src/runtime/app, since its content never changes